File: 0322-coin-change/0322-coin-change.cpp
class Solution {
public:

    int coinChange(vector<int>& coins, int amount) {
        if(amount==0) return 0;
        int n=coins.size();
        int dp[10001] = {0};
        dp[0] = 1;
        for(auto &e: coins) {
            for(int amt=e; amt<amount+1; amt++) {
                int tar = amt-e;
                if(tar==0) dp[amt] = 1;
                else if(dp[tar] && dp[amt]==0)
                    dp[amt] = 1 + dp[tar];
                else if(dp[tar]) dp[amt] = min(1+dp[tar], dp[amt]);
            }
        }
        if(dp[amount]==0) return -1;
        return dp[amount];
    }
};

File: 0322-coin-change/0322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        if (amount == 0)
            return 0;
        int[] dp = new int[10001];
        dp[0] = 1;
        for (int e : coins) {
            for (int curr = e; curr < amount + 1; curr++) {
                int tar = curr - e;
                if (tar == 0) {
                    dp[curr] = 1;
                } else if (dp[tar] != 0) {
                    if (dp[curr] == 0) {
                        dp[curr] = 1 + dp[tar];
                    } else {
                        dp[curr] = Math.min(1 + dp[tar], dp[curr]);
                    }
                }
            }
        }
        return dp[amount] == 0 ? -1 : dp[amount];
    }
}